Refactor <List> to support custom markup

samueljseay opened the pull request:
Fixes woocommerce/woocommerce-admin#6772

The current `<List>` is not suitable for rendering different markup in the list items, which we need for several upcoming stories in https://github.com/woocommerce/woocommerce-admin/issues/6633. Some of the inflexibility also makes it hard to take different kinds of accessibility approaches depending on use as well (as mentioned in woocommerce/woocommerce#32278).

Changes:

1. Add a deprecation warning to the existing `<List>` component
2. Create new components `<ExperimentalList>` and `<ExperimentalListItem>`, these can be consumed once the package is published with the understanding that they're still experimental until `<List>` is removed.


4. See the component props for the full API, but it supports transitions, disabling padding on list items, and passing any custom markup via `children`.

### Accessibility

I haven't tested every combination of attributes, but I have keyboard tested the scenarios around utilising list items as links mainly to fulfil the use cases described by woocommerce/woocommerce#32278. Testing every possible combination seems difficult, but I'm confident almost anything like a `<menu>` would be implementable by using the overrideable props.

Later if we want to create more specific a11y friendly use cases like a `menu` we could create components that derive from `<List>`.

-   [x] I've tested using only a keyboard (no mouse)
-   [x] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
-   [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Detailed test instructions:

1. Probably the simplest way to test this would be to have a go at creating a new `<ExperimentalList>` anywhere in the application and see that the basic styling, markup and aria behaviour work correctly.

Here's some sample code that tests a couple of cases:

```js
import { ExperimentalList, ExperimentalListItem } from '@woocommerce/components';

<ExperimentalList>
	<ExperimentalListItem
		disableGutters
		component="a"
		href="https://example.com"
	>
		<div>Hello world</div>
	</ExperimentalListItem>
	<ExperimentalListItem
		onClick={ () => {
			console.log( 'do something' );
		} }
	>
		<div>Hello world</div>
	</ExperimentalListItem>
	<ExperimentalListItem
		onClick={ () => {
			console.log( 'do something' );
		} }
	>
		<div>Hello world</div>
	</ExperimentalListItem>
</ExperimentalList>
```

2. Check that you see deprecation warnings in wc-admin for the old list.


Other Notes:

* I don't think this warrants test instructions since its a change in packages

No changelog as the package contains a changelog

samueljseay committed:
Create the new experimental list, default to rendering it when items are not supplied.

samueljseay committed:
Remove typings from the List, ListItem.

samueljseay committed:
Add react-transition-group as dependency.

samueljseay committed:
Fix warnings by wrapping with CSS transition, pass down props from List.

samueljseay committed:
Remove type annotations left in list-item.

samueljseay committed:
Duplicate some styles in the higher level woocommerce-list-item element, to make them work in the new list.

samueljseay committed:
Fix accessibility issues, adjust the role logic

samueljseay committed:
Add more testing of the roles logic.

samueljseay committed:
Mark the old list stories as deprecated, add a basic story for the new style.

samueljseay committed:
Expose ListItem without Experimental in the name.

samueljseay commented on bin/packages/build.js:
```diff
@@ -48,7 +48,7 @@ function getPackageName( file ) {
 
 const isJsFile = ( filepath ) => {
-	return /.\.js$/.test( filepath );
+	return /.\.(js|ts|tsx)$/.test( filepath );
```
These bin changes are cherry picked from another PR branch to get the package build working properly. I can discard them before merge.

samueljseay commented on packages/components/package.json:
```diff
@@ -55,7 +55,7 @@
 		"react-dates": "^17.2.0",
 		"react-router-dom": "5.2.0",
-		"react-transition-group": "4.4.1"
+		"react-transition-group": "^4.4.1"
```
This change is probably not needed. I'll remove this.

samueljseay commented on packages/components/src/list/experimental-list.tsx:
```diff
@@ -0,0 +1,50 @@
+			className="woocommerce-list"
+			{ ...otherProps }
+		>
+			{ /* Wrapping all children in a CSS Transition means no invalid props are passed to children and that anything can be animated. */ }
```
Unfortunately *any* direct child of TransitionGroup gets a bunch of props injected to it regardless of element type. react-transition-group is unfortunately not very actively maintained, so I've worked around this by wrapping all children in a CSSTransition, otherwise we get a lot of warnings that we're passing illegal props into the child components.

It means it is possible to animate any child of the list in the same way you can animate a normal list item, I'm not sure thats useful, but until we can move away from react-transition-group I don't see a better way to do this. Maybe we could create a dummy component that swallows the redundant props? I'm open to ideas.

samueljseay commented on packages/components/src/list/index.js:
```diff
@@ -19,37 +21,49 @@ function List( props ) {
-					}
-				);
+	if ( props.items ) {
+		deprecated( 'List.items prop', {
```
warn that this list style is deprecated.

samueljseay committed:
Remove some redundant code changes.

samueljseay commented on packages/components/src/list/style.scss:
```diff
@@ -7,11 +7,21 @@
 	display: flex;
 	align-items: center;
 	margin-bottom: 0;
+	text-decoration: none;
```
I've duplicated some styles that were in list-item-inner up one level here. It doesn't affect existing styling so I think it's fairly low risk.

samueljseay committed:
Fix linting issues

samueljseay committed:
Add types package.

samueljseay committed:
Cast the type of children to TransitionGroup.

samueljseay commented on packages/components/src/list/experimental-list.tsx:
```diff
@@ -0,0 +1,53 @@
+					}
+
+					return child;
+					// TODO - create a less restrictive type definition for children of react-transition-group. React.Children.map seems incompatible with the type expected by `children`.
```
I don't think I can avoid a cast here without changing the type declaration of TransitionGroup. Happy to hear advice on this one. I could also write a type def instead of relying on the definitely typed one. I can see in their type defs they try to avoid children being something incompatible with the props they pass down into transition elements, but I don't think it covers things like text, booleans and nulls as covered by `React.ReactNode`.

samueljseay commented:
@mattsherman you may want to take a look at the ARIA work here if you're interested. I'd be keen for any feedback. The goal has been to make the aria roles more flexible but default to the ones you outlined in #4387

psealock commented on packages/components/src/list/experimental-list-item.tsx:
```diff
@@ -0,0 +1,76 @@
+	const tagRole = role || ( component !== 'a' && hasAction && 'link' ) || '';
+	const TagName = component || 'div';
+	const tagTabIndex = tabIndex || ( hasAction ? '0' : null );
+
```
Something seems a bit off in this logic. Here is my opinion:

* The `tagName` should always be `a` if an `href` is supplied, no need for `component === 'a'`
* The `onClick` should always render a `button` tag, unless there is an `href`. Creating a `div` with role `a` is a roundabout method that prevents the browser from displaying a preview of the link. It also assumes clicking the item results in a navigation. I don't think we should be guiding developers toward navigating in an `onClick`

samueljseay replied:
@psealock I agree with all of this. I went back and forward on these so its great to have some feedback. I'll implement this.

samueljseay replied:
I think this is resolved in https://github.com/woocommerce/woocommerce-admin/pull/6787/files#r614489777 keen to hear thoughts.

psealock reviewed (commented):
Hey Sam, this is looking great. I took it for a spin and will continue to do so tomorrow. 

<img width="785" alt="Screen Shot 2021-04-13 at 4 54 48 PM" src="https://user-images.githubusercontent.com/1922453/114498940-350dba00-9c79-11eb-9c4a-fbd41387d669.png">

I left my initial impressions on the default behaviour. I'm not sure what the use case is for WC Admin, so I played with this in my playground extensions to use it from that perspective.

I understand that all the tags and roles are defaults that can be overridden, but if we publish a `List` component, I think the onus should be on us to handle our own scenario and publish something with the least amount of assumptions built in.

P.S. I love the gutters functionality. Thats always been a difficult one to handle and this solution is very nice.

louwie17 commented on packages/components/src/list/experimental-list.tsx:
```diff
@@ -0,0 +1,53 @@
+			{ /* Wrapping all children in a CSS Transition means no invalid props are passed to children and that anything can be animated. */ }
+			{
+				Children.map( children, ( child ) => {
+					if ( isValidElement( child ) ) {
```
Would you also want to check if the child is a ExperimentalListItem? Given `isValidElement` just checks if it is an element.

samueljseay replied:
@louwie17 not in this case. any child is legal and we have to avoid a div getting injected with the props from TransitionGroup

louwie17 commented on packages/components/src/list/experimental-list.tsx:
```diff
@@ -0,0 +1,53 @@
+								in={ inTransition }
+								enter={ enter }
+								exit={ exit }
+							>
```
Do you think it be worth making the CSSTransition optional? I know I usually have to overwrite the transition style when making use of the component, that is using it outside of the task list context.

samueljseay replied:
@louwie17 thats great feedback and exactly the kind of thing I'm looking for thank you! Yes it sounds like that would be useful. I'll explore that.

samueljseay replied:
@louwie17 I was trying to decide how to move this one forward, I really like your idea here, but also I'm not sure the best way to implement.

If we make it a prop like `disableTransition` then it kinda needs to be on the parent and child, and we conditionally render the css transition wrapper in both. I could live with that I guess although it feels a bit not-right.

One other idea I had that I don't think i could do here, but is worth considering: we could in future iterations, remove the animation css from this package, and then in the one or 2 places where we do use it, we could add it specifically to those packages, because thats kinda how CSSTransition works anyway, its completely opt-in. all it does is add some classes based on state transitions. So in this way it could be optional until you choose to write an animation for it. 

What do you think? I'm open to ideas/thoughts.

louwie17 replied:
Some good suggestions, I like keeping the `CSSTransition` given it will turn messy quickly if we do add conditional logic there. 
In terms of your suggestions I would lean towards number 3, I don't think we need more then one transition option (the current one is probably fine for now), defaulting to that but can be overwritten, we would have to rely on a prop.
It just add's an extra class that includes the transition styles.

I guess something like `disableTransition` might still work better, just removing a `woocommerce-list-transition` class.

I suppose we could also just remove the default transition, but have the css available within a wrapper. Then in the places we want to use it, we just pass the extra className in (we would need to clarify this in the docs quite a bit).

samueljseay replied:
@louwie17 

After some thought on this I have some conclusions:

1. We don't really have many requirements for transitions right now so it has to be kept simple
2. I can imagine though that for a list that transition support is super useful as soon as your list is mutable, so I want to support this for sure and I think long term others will want control over the type of animation.
3. I think by default it should be *not* animated. Many non-mutable lists just won't need it so opting in seems correct.
4. I don't like documenting class names as much as I like documenting prop types. 

So I'm thinking we could start by requiring that users pass a `animation` prop to the component to switch on animations. It could be an enum or a set of consts, but for now we support just one animation. Name of this animation TBD, I think I prefer naming it based on the kind of animation it is rather than calling it something like `default`.

If the user passes the animation prop we will assign the right class name based on the animation type. We can document that we only support one animation for now. I see this as not being perfect, if someone needs a different animation in a pinch they might have to do some class name hacks, but since this is sort of an effort to maintain some backward compatibility with the existing transition, I see this as an ok compromise. If some day we get requests to make animations more customisable or whatever, then we can reassess at that time. I have no idea if this feature will be wanted right now so trying to keep it to supporting what we did before.

jeffstieler commented:
What mechanism are you thinking extension developers will use to add tasks to an extension list? SlotFill or a filter? In relation to [this question](https://github.com/woocommerce/woocommerce-admin/issues/6638#issuecomment-819535183), it may be necessary to have knowledge of the modified extension task array in other areas of the app for the purpose of toggling controls (like the show/hide item in the display options menu).

jeffstieler commented on packages/components/src/list/experimental-list-item.tsx:
```diff
@@ -0,0 +1,76 @@
+				{ ...otherProps }
+				className={ `woocommerce-list__item ${
+					otherProps.onClick ? 'has-action' : ''
+				} ${ className } ${ disableGutters ? '' : 'has-gutters' }` }
```
This looks like a good use case for `classnames`.

samueljseay replied:
I really hate classnames, why use a library to do string concatenation? 😅 I do agree though this has quite a few conditionals in it. Happy to change if its preferred.

samueljseay replied:
I have to confess as much as i hate using libraries it looks nicer with classnames. Fixed in https://github.com/woocommerce/woocommerce-admin/pull/6787/commits/41ba9a8fbe1a9cb478787aacdadf51432e02705f

jeffstieler commented on packages/components/src/list/experimental-list-item.tsx:
```diff
@@ -0,0 +1,76 @@
+};
+
+type ListItemProps = {
+	onClick?: () => void;
```
On "click" seems misleading considering this is attached to `KeyboardEvent`.

samueljseay replied:
I think you've just spotted a mistake in the code. it should be attached to both. Attaching it to keyboard is to satisfy keyboard nav requirements but it should be attached as the onClick handler as well. thanks for spotting this, I should be able to cover this with a test too.

samueljseay replied:
I ended up removing this from being declared directly in the props since now I'm defaulting to acting like a link. now onClick is attached as part of spreading `otherProps`

mattsherman commented:
> @mattsherman you may want to take a look at the ARIA work here if you're interested. I'd be keen for any feedback. The goal has been to make the aria roles more flexible but default to the ones you outlined in #4387

@samueljseay Thanks for the ping, and thanks for keeping accessibility in mind! I ran out of time to look at this in detail today, but will do so tomorrow. It sounds like the approach you are taking is reasonable. I'd probably ideally want specific "types" of List that handle the different UX cases without the implementor needing to think too much about roles or such.... `NavigationList`, `Menu`(ah, I see you mentioned that... "Later if we want to create more specific a11y friendly use cases like a menu we could create components that derive from <List>.").

Do we have any expected usage of this new List component that isn't nav?

samueljseay commented:
> Do we have any expected usage of this new List component that isn't nav?

@mattsherman I kinda don't want to frame it that way, because this is part of our components package which is published for anyone outside our team to use. We don't know if other teams have been using it for non-nav purposes without trying to somehow survey users of the package. Because the existing component was rendered as a menu, maybe someone was using it that way. 

Being that this component is called `List` and not I dunno, `NavList` or something, I think it makes sense for it to be a design component that gives you the tools to build any kind of list-like feature that fits the woo look and feel, which means it could totally be valid as a menu, list of links, or just a list of items.

Making roles customisable means you can implement any a11y scenario you can consider, but as you said, if we wanted to provide an off the shelf menu later, we can build it on top of this component too. I see it as the best of both worlds.

mattsherman commented:
> > Do we have any expected usage of this new List component that isn't nav?
> 
> @mattsherman I kinda don't want to frame it that way, because this is part of our components package which is published for anyone outside our team to use. We don't know if other teams have been using it for non-nav purposes without trying to somehow survey users of the package. Because the existing component was rendered as a menu, maybe someone was using it that way.

Fair point -- I forget that others could already be using it for non-nav purposes.

mattsherman commented:
I gave this a brief spin today and things do not act how I expected they would...

JSX source:

```jsx
<List>
    <ListItem href="https://woocommerce.com">
	<div>WooCommerce.com</div>
    </ListItem>
    <ListItem href="https://wordpress.com">
	<div>WordPress.com</div>
    </ListItem>
    <ListItem href="https://google.com" component="a">
	<div>Google</div>
    </ListItem>
</List>
```

HTML output:

```html
<nav class="woocommerce-list">
    <div role="" href="https://woocommerce.com" class="woocommerce-list__item   has-gutters">. 
       <div>WooCommerce.com</div>
    </div>
    <div role="" href="https://wordpress.com" class="woocommerce-list__item   has-gutters">
        <div>WordPress.com</div>
    </div>
    <a role="" href="https://google.com" class="woocommerce-list__item   has-gutters">
        <div>Google</div>
    </a>
</nav>
```

I expected:
- That `role` would be set to `link` on the `ListItem` divs that didn't have a `component` specified, since that's what the description of the PR said
    - I also expected, though this wasn't ever specified, that if I created a `ListItem` with an `href`, the resulting output would be some sort of clickable element that navigated to the supplied `href` without me having to implement anything
- That `role` would not be present at all on the `a` element

samueljseay commented:
@mattsherman Thanks looks like a couple little bugs there. I hadn't considered also that React would still put the prop there if it was empty. I'll fix these and add some tests. I also intend to change the behavior a bit based on these suggestions: https://github.com/woocommerce/woocommerce-admin/pull/6787#discussion_r612127780

samueljseay committed:
Adjust the list to assume a purpose of being a list of links.

samueljseay commented on packages/components/src/list/experimental-list-item.tsx:
```diff
@@ -0,0 +1,78 @@
+	// Everything else you might pass into an HTML element
+	...otherProps
+} ): JSX.Element => {
+	// for styling purposes only
```
@psealock @mattsherman wondering if you could take another look. Looking back on my first approach it was pretty naive, some of the assumptions made no sense based on how I implemented the `<List>`

Let me talk through reasoning of what I have now.

I originally made the default component for List a `nav` to satisfy the original problem of the list not being semantically correct or satisfying aria concerns raised by @mattsherman , but then I realised thanks to @psealock 's comments that all the work I was doing in the list item, was circumventing it acting like a link, and there was convoluted logic to add roles. It was silly on retrospect, if the default is a link list, then let's default to an actual link! roles after all are a workaround for when you *can't* use the semantically correct markup. But here we totally can.

So the approach is now: List is a `nav`, list items are `a` tags. no roles needed since its all semantically correct. link previews work. 

If you want to introduce a different semantic approach you can do that by overriding `component` in List and ListItem then you can do that. If you provide an `onClick` handler and a different `component` then it'll not show a link preview but it will have a `cursor: pointer`, so its purely presentational, providing the right roles will be up to the implementer.

Keen to hear your thoughts! One other upside I can think of is, that it does satisfy the core use case within wc-admin but without painting anyone else using it into a corner either.

samueljseay replied:
Also worth noting I never bothered with deciding whether to render a `button`, that one thing feels like the wrong approach to me. Button has a lot of styling associated with it which is a pain, I think we should leave it up to the implementer if they want to use a button or whether they want to add a role to a div.

samueljseay commented on packages/components/src/list/style.scss:
```diff
@@ -3,15 +3,29 @@
 	padding: 0;
 }
 
+a.woocommerce-list__item {
```
Since the default purpose is a link, this is some helper styling to stop the inner content from being made link color blue by default.

samueljseay committed:
Adjust the tests for the new behaviour.

samueljseay committed:
Fix css mangled by prettier.

louwie17 commented on packages/components/src/list/experimental-list-item.tsx:
```diff
@@ -0,0 +1,78 @@
+
+type ListItemProps = {
+	// control whether to display padding on list item or not.
+	disableGutters: boolean;
```
@samueljseay `disableGutters` should also be optional here, given it defaults to `false`.

louwie17 replied:
I suppose the same for the `CssTransitionProps`

samueljseay replied:
Thanks will fix the optionals

samueljseay replied:
Fixed here https://github.com/woocommerce/woocommerce-admin/pull/6787/commits/6d47d9a6fde9cf27811c6a72897e82817837b759

samueljseay committed:
Fix types that are optional.

psealock commented:
Nice @samueljseay, this is great

Since the assumptions built into this component are geared towards a list of links, I'd like to step back and ask what else this component could be used for?

Seeing the gutter functionality and layout issues taken care of, I would be tempted to use this component as a card that houses any list of items, UI pieces, or links. Here is something I came across today that could fit the bill:

<img width="578" alt="Screen Shot 2021-04-19 at 1 05 20 PM" src="https://user-images.githubusercontent.com/1922453/115168911-aa153f80-a110-11eb-812b-cd3ed6ffc352.png">

With this `<List />` component, I don't need to worry about any CSS layout at all and focus on content. However, if I tried to use this component as is, I would end up with some very odd markup with buttons and toggles inside of `<a/>` tags.

I think our use case as it is, a `nav` collection of links, is the use case that requires modifying from a sane default. Simply rendering a list of things to display should be the base line use case for maximum usability. We shouldn't ask devs to supply a prop to disable built in functionality just to render a `div`.

Also, it is more semantically correct to have items in a `ul`. https://developer.mozilla.org/en-US/docs/Web/HTML/Element/nav

```html
<nav class="menu">
  <ul>
    <li><a href="#">Home</a></li>
    <li><a href="#">About</a></li>
    <li><a href="#">Contact</a></li>
  </ul>
</nav>
```
Of course, if the defaults here are something we'd like to keep, we should rename the component to reflect that.

samueljseay commented:
@psealock This was intended to present backwards compatibility with the existing component. The idea was that people using it currently would not have to do anything until they're ready to transition across, so I've kept it inline with its original purpose as the default, but the more I think about it inline with your comments **maybe it makes no sense at all**. We don't really improve the existing component enough to make it useful in a broader context in this PR.

So here's a proposal of changes I'd introduce instead:

1. deprecate the old list.
2. complete departure from old component, create a new list that is completely independent of concepts like lists of links, use a `ul` or `ol` and `li`.
3. we can either: create a new component like `NavList` in `packages` that consumers the new `List` or we can do it internally in wc-admin if we're not sure what we want to expose going forward
4. Depending on what we do in `3` we could either have the deprecation warning tell you to use `List` to mimick the capabilities or `NavList` to replace usage.

psealock commented:
>So here's a proposal of changes I'd introduce instead:

❤️ every bit of this! I didn't realize there were backwards compatibility issues involved, but I think you are right that breaking free to start fresh makes the most sense for everyone

samueljseay committed:
Rework the list item.

mattsherman commented:
> So here's a proposal of changes I'd introduce instead

I think this proposal will result in component that is more easily used in more contexts. Makes sense.

samueljseay committed:
Export new list as experimental, fix accessibility styles and deprecation notice.

samueljseay committed:
Make the basic transition opt-in.

samueljseay commented:
@mattsherman @psealock @louwie17 keen for ya'all to take another look at the new iteration (no pressure I know I've taken a lot of your time already). I've been playing with it a bit locally and feeling like its much more flexible now. 

@louwie17 the transition is opt in now and can even be applied on individual list items if you only want to animate some of the list. I played around with it a bit and this felt pretty intuitive, keen for you to have a play with that.

@psealock @mattsherman I had a little experimental try at implementing a nav link list where we specifically wanted the semantics of a `ul` with `li` containing `a`, and it worked quite well, but it wasn't a matter of simply overriding `component`. Here's an example:

```js
const Component = () => {
		const dummyList = [
			{ title: 'Hello world 1', id: 1 },
			{ title: 'Hello world 1', id: 2 },
			{ title: 'Hello world 1', id: 3 },
		];

		return (
			<nav>
				<ExperimentalList>
					{ dummyList.map( ( { id, title } ) => {
						return (
							<a href="https://foo.com" key={ id }>
								<ExperimentalListItem>
									<div>{ title }</div>
								</ExperimentalListItem>
							</a>
						);
					} ) }
				</ExperimentalList>
			</nav>
		);
	};

```

So I'd say its flexible enough and we could easily create some nav purpose specific components that wrap the list to provide this for people as a separate component.

If ya'all are happy I'll write the tests and adjust the stories but just keen for some feedback before I go further.

louwie17 commented on packages/components/src/list/style.scss:
```diff
@@ -26,6 +26,46 @@ a.woocommerce-list__item {
+		&.exit,
+		&.exit-active {
+			transition: none;
+			transform: none;
```
Hmm instead of overwriting the transition properties when it is disabled, would it be worth to wrap the transitions instead with either `:not(.transition-disabled)` or something specific to the transition `.list-slide-right`?

samueljseay replied:
I think I did it this way to not impact the existing component, but afaict transitions on the old one may have been broken anyway because when react-transition-group was updated the format of the expected classes for transitions changed. I'll explore your suggestions here 👍

samueljseay replied:
Addressed here: https://github.com/woocommerce/woocommerce-admin/pull/6787/commits/5d04cb4baa82c80294c58ba6a591d27bdca3e319

louwie17 commented on packages/components/src/list/style.scss:
```diff
@@ -103,7 +143,7 @@ a.woocommerce-list__item {
 	&.is-complete {
 		.woocommerce-task__icon {
-			background-color: var(--wp-admin-theme-color);
+			background-color: var( --wp-admin-theme-color );
```
My editor does this as well, adding spaces on auto format, but it appears our linter doesn't like this and prefers no spaces.
It be nice to get to a consensus, between the two, I guess it hasn't quite driven me crazy enough to update it yet (just throwing it out, no need to fix that as part of this).

samueljseay replied:
@louwie17 whoops, this will fail the css lint anyway. I'll fix it. We could add a prettierignore but, actually we should just add some prettier option overrides for the bracket spacing because I *think* thats the only conflict with stylelint

louwie17 reviewed (commented):
I think this looks great @samueljseay, thanks for sticking with it through the many iterations, this looks very flexible now, and I haven't run into any issues (aside from the optional params) when updating task-list.

I did offer one suggestion with the transition css, hoping to prevent overwriting css attributes if possible.

mattsherman commented:
> @mattsherman @psealock @louwie17 keen for ya'all to take another look at the new iteration (no pressure I know I've taken a lot of your time already). I've been playing with it a bit locally and feeling like its much more flexible now.

I'll take a closer look tomorrow!

> @psealock @mattsherman I had a little experimental try at implementing a nav link list where we specifically wanted the semantics of a `ul` with `li` containing `a`, and it worked quite well, but it wasn't a matter of simply overriding `component`. Here's an example:
> 
> ```js
> import {
> 	ExperimentalList,
> 	ExperimentalListItem,
> } from '@woocommerce/components';
> 
> const Component = () => {
> 		const dummyList = [
> 			{ title: 'Hello world 1', id: 1 },
> 			{ title: 'Hello world 1', id: 2 },
> 			{ title: 'Hello world 1', id: 3 },
> 		];
> 
> 		return (
> 			<nav>
> 				<ExperimentalList>
> 					{ dummyList.map( ( { id, title } ) => {
> 						return (
> 							<a href="https://foo.com" key={ id }>
> 								<ExperimentalListItem>
> 									<div>{ title }</div>
> 								</ExperimentalListItem>
> 							</a>
> 						);
> 					} ) }
> 				</ExperimentalList>
> 			</nav>
> 		);
> 	};
> ```

It seems odd to be wrapping the `ExperimentalListItem` in an `a`. Two thoughts...

1) Why even have `ExperimentalListItem` at all in this case? or...
2) Why not wrap things the other way...

```jsx
<ExperimentalListItem key={ id}>
    <a href="https://foo.com">{ title }</a>
</ExperimentalListItem>
```

3) Why not be able to do something like...

```jsx
<ExperimentalListItem tag="a" href="https://foo.com" key={ id }>{ title } </ExperimentalListItem>
```

The larger question... in this nav case, what is `ExperimentalList` and `ExperimentalListItem` really providing to the implementor?

louwie17 commented on package.json:
```diff
@@ -153,12 +153,14 @@
 		"@testing-library/react": "11.2.6",
 		"@testing-library/react-hooks": "3.7.0",
 		"@testing-library/user-event": "12.8.3",
+		"@types/classnames": "^2.2.11",
```
FYI @samueljseay this package doesn't contain any types, and the latest ~2.3.1 version of classnames comes with types: https://github.com/JedWatson/classnames/blob/master/HISTORY.md (just realized this myself).

samueljseay replied:
oh awesome. ok, thanks I'll fix this.

samueljseay replied:
Fixed in https://github.com/woocommerce/woocommerce-admin/pull/6787/commits/f755dd4696491bf008d20f3958fe432e4b553d49

samueljseay commented:
> Why even have ExperimentalListItem at all in this case? or...

@mattsherman in the case of the nav list the feature is adding styling support, things like transitions etc. This thing is a design element at the end of the day thats meant to be built on top of to build other things

> Why not wrap things the other way...

I agree the `a` as child of the ListItem is better, it just needs some style adjustments to make that work the same way the old one did, I was trying to reproduce existing behaviour. 

> Why not be able to do something like...
`<ExperimentalListItem tag="a" href="https://foo.com" key={ id }>{ title } </ExperimentalListItem>`

It would generate this structure:

`<ul><a>title</a></ul>`

its not really semantically correct for a list to have items that are not in lis. But then thats a similar problem with wrapping in an `a` tag.

mattsherman commented:
> > Why even have ExperimentalListItem at all in this case? or...
> 
> @mattsherman in the case of the nav list this list is adding styling support, things like transitions, tab indexing items, etc. This thing is a design element at the end of the day thats meant to be built on top of to build other things.

Yes, that is true. Thinking about it more I do agree that it provides enough functionality to make it useful to build on as a base in that scenario.

> > Why not be able to do something like...
> > `<ExperimentalListItem tag="a" href="https://foo.com" key={ id }>{ title } </ExperimentalListItem>`
> 
> It would generate this structure:
> 
> `<ul><a>title</a></ul>`
> 
> its not really semantically correct for a list to have items that are not in lis.

Oh, of course. Oops! Ignore me. 😄

> But then thats a similar problem with wrapping in an `a` tag, I wasn't sure if it was semantically ok for a li to be wrapped in a link. 

> Either way the component is very flexible for any outcome we want I think.
>
> I think overall I was just trying to prove this is flexible enough to be reused in future components such as a nav list, and I don't think it makes sense to copy paste a lot of style, logic etc to build that again from scratch.

Very true — I agree! Thanks for keeping future use cases such as this in mind from the start!

samueljseay committed:
Update classnames to get types.

samueljseay committed:
Add tests, adjust behaviour for css disable.

samueljseay committed:
Fix css lint, remove types/classnames, update tetsing-library/user-event.

samueljseay committed:
Fix the basic story for ExperimentalList.

samueljseay commented:
@psealock or @louwie17 I think I've addressed the feedback so far. There are some issues with npm install on CI, which I'll look at but I've updated the stories. last thing to do is add a changelog entry to packages.

psealock reviewed (commented):
I'm like the direction this is going @samueljseay!

I see two problems with the resulting markup using the code sample in the original comment:

<img width="424" alt="Screen Shot 2021-04-22 at 5 10 09 PM" src="https://user-images.githubusercontent.com/1922453/115658815-a04c4000-a38d-11eb-932b-cd885a583174.png">

First, an `<a/>` can't be the child of a `<ul/>`. I think the logic could include the detection of an `href` and wrap in an `<li/>` before rendering the anchor tag. 

I'll also note that my personal preference is `component="a"` is not required. It can be inferred from the presence of an `href`. I understand thats just me, so not a blocker.

Secondly, the resulting markup of an `onClick` prop on a `div` violates a11y. If I were to simply write that html, our linter has several complaints, and for good reason:

<img width="1146" alt="Screen Shot 2021-04-22 at 4 50 41 PM" src="https://user-images.githubusercontent.com/1922453/115658295-aaba0a00-a38c-11eb-9aa9-53d8eac5f8dc.png">

Keyboard users don't have the ability to fire the `onClick`. I see two ways to deal with this:

1. Render a `<button />` tag if an `onClick` is present and remove the default styling.
2. Apply a `role="button"` to the div. This also must include [at least one keyboard listener](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/master/docs/rules/click-events-have-key-events.md) according to our linting rules, as seen in the screenshot above. So it will need to look something more like this:
```html
<div
	role="button"
	tabIndex="0"
	onClick={ () => {} }
	onKeyPress={ () => {} } //This will need to mimic the same behaviour as onClick for "Enter"
>
	hello
</div>
```

samueljseay commented:
@psealock sorry for the confusion about the original sample I should have changed it after discussion with Matt. The better approach would be to have an a tag as child of the list item. We can achieve having a fully clickable list item by removing padding via disableGutters. We don't need the list item itself to directly be clickable in that case.

I don't think it make sense to try use buttons or infer things from href in this list component. And I agree overriding with "a" was simply incorrect in my original example

So basically my original example isn't very useful.

samueljseay commented:
@psealock forgot to address onClick. Not sure what the issue is there? It's not semantic without a role on it. But everything is over rideable so if you decide to implement a menu with this you can add the role that makes sense but we are no longer painting this to be used for that. It's just a plain list that can become that with the adjustment of props. 

I think we can't have a super flexible thing *and* a thing that makes a bunch of assumptions. There is no specified API to turn this into a nav list anymore just a component that is flexible enough that we could implement that later if needed

louwie17 commented:
>  There are some issues with npm install on CI, which I'll look at but I've updated the stories. last thing to do is add a changelog entry to packages.

@samueljseay this should get fixed if you get the latest package-lock now from `main` and do a fresh install. If it worked correctly the package-lock should have `file:packages/components` for the component version, etc...

mattsherman commented:
@psealock 

> Secondly, the resulting markup of an `onClick` prop on a `div` violates a11y. If I were to simply write that html, our linter has several complaints, and for good reason:

@samueljseay 

> @psealock forgot to address onClick. Not sure what the issue is there? It's not semantic without a role on it. But everything is over rideable so if you decide to implement a menu with this you can add the role that makes sense but we are no longer painting this to be used for that. It's just a plain list that can become that with the adjustment of props.

Ideally, we shouldn't generate HTML that isn't accessible, by default. But, I'm not sure if we should generate keyboard handlers automatically either. If we want it to act like a `<button>` by default when an 'onClick' is provided, it would seem like it would best to just render `<button>`. Supporting `onClick` on the list item itself is just one way a consumer can use this. If they want something more custom, they can hook up whatever handlers they want in the children of the list item.

samueljseay commented:
@mattsherman onClick can be passed to any jsx element. We don't really do anything to support beyond adding a style that makes it look clickable if you choose to. We could probably even remove that if that's the tension here. 

We are not at any point generating a11y non compliant structure any more than someone who renders a div with onClick and chooses not to put a role on it. I've ditched the approach of trying to make this magically support link lists here. 

By passing through jsx compliant props we are doing what any component library would do but when someone builds something other than a basic list the a11y falls to them. 

I'm really not understanding the arguments at this stage.

mattsherman commented:
> We are not at any point generating a11y non compliant structure any more than someone who renders a div with onClick and chooses not to put a role on it.

That’s an excellent point.

> I'm really not understanding the arguments at this stage.

I guess it just comes down to how much we want the component to help implementors out with accessibility. Perhaps it is best to not get fancy and leave it up to them. As you say, it’s no different than if they were not using the component.

I trust your approach to reusability when creating a component, and feel like I’m getting a little hung up still on the idea that the component should be “doing more” for the consumer. :-) I think the “minimal” approach will ultimately provide the biggest payoff in ultimate reusability though!

samueljseay commented:
>  and feel like I’m getting a little hung up still on the idea that the component should be “doing more” for the consumer.
@mattsherman I think this is valid, I think we can build more components on top of these that do more as we go forward. We don't have a strong direction for the component lib right now but I think some of the work we're doing in wc pay should highlight some opportunities to provide "doing more" components.

samueljseay committed:
Conditionally add keyDown and role based on onClick handler. Adjust tests to new behaviour.

samueljseay commented:
Based on discussion with @psealock I made some changes:

I removed the ability to override the component on the list item, I kept the override on the list but renamed it to `listType` and with typescript the type gives a soft restriction to use `ol` or `ul` to keep it semantic, this is after all a `list`. 

When you add an onClick to an item which is fairly valid if the role is right, I've added the a role of `button` and kept the keyDown handler. I think this makes it really easy and convenient to add basic keyboard navigation and it keeps the semantics right without much effort from the consumer.

I think these are pretty good compromises. Not too many assumptions made, but also enough thought put in for it to be useful.

The last thing @psealock and I discussed was moving these new components to the `@woocommerce/experimental` package which I think is a great idea. If we do this, then other plugins like wc pay can consume these new components without bundling 2mb of `@woocommerce/components`. I haven't done this work yet, if I don't get to it, I'd love if someone on @woocommerce/ghidorah could pick it up.

psealock committed:
tabIndex only on hasAction

psealock reviewed (approved):
This is looking good. I added one change in https://github.com/woocommerce/woocommerce-admin/pull/6787/commits/d05689d2477be12259152317afeeb4cebd11ab83 which only adds the `tabindex="0"` attribute if the list item `hasAction`, ie an `onClick` handler is present. This makes only list items with click handlers (and keyDown) tabbable. Items inside list items, like `<a/>` are still tabbable.

>moving these new components to the @woocommerce/experimental

I think this is a good idea because it can be bundled independently of `@woocommerce/components`. I also think its a good idea to follow this up in another PR because it sets a precedent and it would be a good opportunity to make sure we have a good idea of how it gets integrated back and handles deprecation/backwards-compat gracefully.

I'm approving, but will give someone else a chance to review because I authored a small portion of the PR

louwie17 committed:
Update the package-lock

louwie17 committed:
Add changelog

louwie17 reviewed (approved):
Approving this as well, I think things are looking pretty polished now, and given it doesn't replace the List component yet, it allows for future (non breaking) changes if necessary.

Thanks for all your hard work on this @samueljseay, and sticking with it, through all the different iterations 💯

louwie17 committed:
Fix transition styling
